doc/new-guidelines.md

```diff
@@ -74,6 +74,32 @@ This mistake mostly happens with functional components.
 
 It's important to name functional components with `f-` prefix.
 
+### Default value when destructuring
+
+Too often callers pass nil values because values can be wrapped in a `when` for example.
+In this case, the default value is not applied, because :or macro will use default only when the value is absent.
+Instead, use `(or value some-default-value)` in a `let` expression or as a parameter value.
+
+```clojure
+;; bad (unreliable)
+(defn- view-internal
+  [{:keys [auto-focus?
+           init-value
+           return-key-type]
+    :or   {auto-focus?     false
+           init-value      0
+           return-key-type :done}}])
+
+;; good
+(defn- view-internal
+  [{:keys [auto-focus?
+           init-value
+           return-key-type]}])
+    (let [auto-focus? (or auto-focus? false)
+          init-value (or init-value 0)
+          return-key-type (or return-key-type :done)])
```

Comment:
I don't fully agree with this approach.

I understand the problems it's creating, but it doesn't look clean, we are also suggesting to shadow vars.

I think this technique should be followed only sometimes, when needed, but these guidelines are thought to be applied everywhere and the code will become (even) more verbose.

to avoid the shadowing, I see cleaner:
```clojure
(defn view [{default-auto-focus? :auto-focus?
             default-init-value  :init-value}]
  (let [auto-focus? (or default-auto-focus? 20)
        init-value? (or default-init-value 0)]))
```

Reply:
Nice suggestion @ulisesmac. I prefer that way too without shadowing.

> I think this technique should be followed only sometimes, when needed, but these guidelines are thought to be applied everywhere and the code will become (even) more verbose.

I agree with you. I think we should adjust the wording in the guidelines. Not all guidelines should be followed 100% of the time. This one I think should be considered by all developers. I personally try to never rely on default destructuring in quo components. Too often I've seen arguments been passed as nil. And I believe a good proportion of the dev team doesn't know nil will overrule default destructured values.

Even with Malli it doesn't help much because we're defining schemas for components with heavy usage of the `:maybe` schema, so nils can still flow in.

Searching by usages of `:or`, I see many have fragile code, waiting to cause small UI bugs or even exceptions, like here (if `:size` is passed as nil an exception will be thrown) https://github.com/status-im/status-mobile/blob/07feff60ec08645da78e3507070594271e87c9d3/src/quo/components/tags/permission_tag.cljs#L137-L151

Maybe this guideline should be considered mostly for quo components? So it could be moved to `quo/README.md`.

Reply:
> Maybe this guideline should be considered mostly for quo components?

Yeah, that in fix the wording will make it better :)

Reply:
I've moved the guideline to the `quo/Readme.md`, please, take a look.

Also, I don't prefer the naming with the `default-` prefix, as the actual default will be the second prop to `or` fn, ex. `20` in `(or default-auto-focus? 20)`. For me, the better naming can be smthng like `param`-auto-focus? or `external`-auto-focus?. Probably, shorter is better.